Site Migration: Don't retry to install migration plugin if it's already installed

nelsonec87 opened the pull request:
closes to  #90978

## Proposed Changes

* Fix an issue on wpcom HTTP client when it is running inside nodejs (test env context). The original version was not enable us to capture errors when we ran the code on the test suite + http_envelop=1.
* Update plugin installation process to assume as a success if the endpoint returns the error `PluginAlreadyInstalledError`

## Testing Instructions
* Checkout this branch in your local env
* Start a migration `/setup/migration-signup` 
* Open the developer tools > network
* Go to the instructions page
* Fix any jetpack connection issue to enable the plugin installation if need
* Wait until the plugin installation is completed. 
* Go to your site and disable the `migrate-guru` plugin
* Apply this [patch] (https://github.com/user-attachments/files/16381662/force_install.patch) in your local env to force the code to try to install the plugin again. `git apply force_install.path`. 
* Reload the instructions page and check if there is only one call to `/sites/${ siteId }/plugins/${ pluginSlug }/install`.
* The endpoint will return the error `PluginAlreadyInstalledError` 
* The UI should not retry the request and assume it was installed correctly.
* The UI should activate the plugin after trying to install it.  

> [!NOTE]  
> The plugin activation endpoint is idempotent. The API doesn't return errors if we try to active the same plugin multiple times.


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

nelsonec87 committed:
Don't try to install plugin if it's already installed

nelsonec87 committed:
Updated tests

gabrielcaires commented on client/landing/stepper/hooks/use-plugin-auto-installation.ts:
```diff
@@ -64,7 +64,14 @@ const usePluginInstallation = ( pluginSlug: string, siteId?: number, options?: O
+				if ( e?.error === 'plugin_already_installed' ) {
+					return false;
+				}
+				return failCount < DEFAULT_RETRY;
```
```suggestion
				return failCount < ( options?.retry ?? DEFAULT_RETRY );
```

gabrielcaires commented on client/landing/stepper/hooks/use-plugin-auto-installation.ts:
```diff
@@ -64,7 +64,14 @@ const usePluginInstallation = ( pluginSlug: string, siteId?: number, options?: O
 		mutationKey: [ 'onboarding-site-plugin-installation', siteId, pluginSlug ],
 		mutationFn: async () => installPlugin( siteId!, pluginSlug ),
-		retry: options?.retry ?? DEFAULT_RETRY,
+		retry:
```
```suggestion
		retry: (failCount, error)
```

nelsonec87 committed:
wip

gabrielcaires committed:
Use the query param properly when request is happening inside node

Writing automated tests I found a case where we are not able to capture the error name because the wpcom-xhr-request lib was never recognizing the request as an enveloped request.
It was generting incosistences between the tests and the real code execution.

gabrielcaires committed:
Improve implementation and tests to follow the rest convention

gabrielcaires committed:
Fix tests

gabrielcaires committed:
Adjust logic to active plugin

markbiek commented:
Looks like that unit test failure is real.

gabrielcaires committed:
Refactor to improve readability

markbiek commented:
Not a blocker to testing but I wanted to note that `git patch fail.patch` doesn't work for me (git v2.45.2).

I also tried `git apply fail.patch` but got the following error:

```
error: patch failed: client/landing/stepper/hooks/use-plugin-auto-installation.ts:91
error: client/landing/stepper/hooks/use-plugin-auto-installation.ts: patch does not apply
```

markbiek reviewed (commented):
@gabrielcaires Is it possible there is a missing commit or the test instructions need to be updated?

(note that `git patch fail.patch` doesn't work for me (git v2.45.2))

I also tried `git apply fail.patch` but got the following error:

```
error: patch failed: client/landing/stepper/hooks/use-plugin-auto-installation.ts:91
error: client/landing/stepper/hooks/use-plugin-auto-installation.ts: patch does not apply
```

I went to make the change manually and I don't see the line `const skipped: SkipStatus` in `client/landing/stepper/hooks/use-plugin-auto-installation.ts` and I see in the changed files of this PR that it was removed.

markbiek reviewed (approved):
Working as expected!
